src/build.d

```diff
@@ -0,0 +1,1131 @@
+        ["git", "describe", "--dirty"].runCanThrow.output.toFile(versionFile);
+    };
+    Rule versionRule = {
+        target: versionFile,
+        commandFun: commandFun,
+    };
+    const sysconfDirFile = env["G"].buildPath("SYSCONFDIR.imp");
+    commandFun = (){
+        env["SYSCONFDIR"].toFile(sysconfDirFile);
+    };
+    Rule sysconfDirRule = {
+        target: sysconfDirFile,
+        commandFun: commandFun,
+    };
+    return [versionRule, sysconfDirRule];
+}
+
+/**
+Main build routine for the DMD compiler.
+Defines the required order for the build rules, runs all these dependency rules
+and afterwards builds the DMD compiler.
+*/
+auto buildDMD()
+{
+    // The string files are required by most targets
+    Rule[] rules = buildStringFiles();
+    foreach (rule; rules.parallel(1))
+        rule.run;
+
+    rules = [lexer, newDelete, dmdConf];
+    foreach (ref rule; rules.parallel(1))
+        rule.run;
+    auto backend = buildBackend();
+
+    // Main DMD build rule
+    Rule rule = {
+        // newdelete.o + lexer.a + backend.a
+        sources: sources.dmd.chain(sources.root, rules[0].targets, rules[1].targets, backend.targets).array,
+        target: env["G"].buildPath("dmd").exeName,
+        command: [
+            env["HOST_DMD_RUN"],
+            "-of$@",
+            "-vtls",
+            "-J"~env["G"],
+            "-J../res",
+            "-L-lstdc++",
+        ].chain(flags["DFLAGS"], "$<".only).array
+    };
+    rule.run;
+}
+
+/**
+Goes through the target list and replaces short-hand targets with their expanded version.
+Special targets:
+- clean -> removes generated directory + immediately stops the builder
+*/
+auto predefinedTargets(string[] targets)
+{
+    Appender!(string[]) newTargets;
+    foreach (t; targets)
+    {
+        t = t.buildNormalizedPath; // remove trailing slashes
+        switch (t)
+        {
+            case "auto-tester-build":
+                "TODO: auto-tester-all".writeln; // TODO
+                break;
+
+            case "toolchain-info":
+                "TODO: info".writeln; // TODO
+                break;
+
+            case "unittest":
+                "TODO: unittest".writeln; // TODO
+                break;
+
+            case "cxx-unittest":
+                "TODO: cxx-unittest".writeln; // TODO
+                break;
+
+            case "check-examples":
+                "TODO: cxx-unittest".writeln; // TODO
+                break;
+
+            case "build-examples":
+                "TODO: build-examples".writeln; // TODO
+                break;
+
+            case "checkwhitespace":
+                "TODO: checkwhitespace".writeln; // TODO
+                break;
+
+            case "html":
+                "TODO: html".writeln; // TODO
+                break;
+
+            case "install":
+                "TODO: install".writeln; // TODO
+                break;
+
+            case "man":
+                "TODO: man".writeln; // TODO
+                break;
```

Comment:
All these targets are nice-to-have (as we eventually want to get rid of the Makefile), but obviously not required for building dmd. I added them here, s.t. adding them doesn't get forgotten.